src/stash/src/transaction.rs

```diff
@@ -179,22 +186,42 @@ impl<'a> Transaction<'a> {
 
     #[tracing::instrument(level = "debug", skip(self))]
     pub async fn upper(&self, collection_id: Id) -> Result<Antichain<Timestamp>, StashError> {
+        // We can't use .entry here because that would require holding the
+        // MutexGuard across the .await.
+        if let Some(entry) = self.uppers.lock().unwrap().get(&collection_id) {
+            return Ok(entry.clone());
+        }
         let upper: Option<Timestamp> = self
             .client
             .query_one(self.stmts.upper(), &[&collection_id])
             .await?
             .get("upper");
-        Ok(Antichain::from_iter(upper))
+        let upper = Antichain::from_iter(upper);
+        self.uppers
```

Comment:
I believe there's a race here: because we run things in various futures (and hence the need for the arc mut), another statement could have committed a new upper after the above await but before the hashmap is updated. I think all the hashmap mutation methods need to only insert a new value if it's greater than the current value to avoid going backward.

Reply:
> I think all the hashmap mutation methods need to only insert a new value if it's greater than the current value to avoid going backward.

That sounds like a good idea to me.

Reply:
I've also realized this was possible in the old cache too, so it's at least not a new behavior.

Reply:
Done.